### ISSUE TITLE
Incorrect `unreachable_pub` warning

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#92903]
I tried this code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=028239a954b1baf0326d31262947a247

```rust
#![warn(unreachable_pub)]
use inner::FlowData;

mod inner {
    pub struct FlowData {
        pub data: Vec<u8>,
    }
}

pub fn f() -> tokio::sync::mpsc::Receiver<FlowData> {
    todo!()
}

async fn foo() {
    println!("{:?}", f().recv().await.unwrap().data);
}
```

I expected to see this happen: No unreachable_pub lint. `foo` only uses publicly accessible APIs, it never mentions `inner`.

Instead, this happened:
```
warning: unreachable `pub` field
 --> src/lib.rs:6:9
  |
6 |         pub data: Vec<u8>,
  |         ---^^^^^^^^^^^^^^
  |         |
  |         help: consider restricting its visibility: `pub(crate)`
  |
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(unreachable_pub)]
  |         ^^^^^^^^^^^^^^^
```

cc @petrochenkov

### Meta

`rustc --version --verbose`: 1.60.0-nightly (2022-01-13 22e491ac7ed454d34669)

**Comment by jyn514** [comment]
In particular, it seems wrong to say that a public field is unreachable whenever the type it belongs to is publicly reachable.

**Comment by tmiasko** [comment]
This was fixed in #92781 (not available as nightly yet).

**Comment by jyn514** [comment]
Yes it was, thank you!
